### DISCUSSION TITLE
[Feature Request] Set theoretic `beartype.is_subtypes()` API for deciding type hint A ⊆ B "compatibility" relations

### DISCUSSION THREAD

**Discussion by tlambert03** [beartype/beartype#132] [Q&A]
First, thanks for this amazing library.  The scope and detail are beary impressive. 🐻

Apologies if I've missed an obvious answer (or previous discussion) of this question: I've been digging through the readme and source code for a couple days but still haven't hit on it, and I'm not sure if this is just fundamentally out of the primary goal here of `isinstance(some_object, some_typehint)`.

I'm looking for something that would tell me whether type hint B is "compatible" with type hint A (something akin to pytypes `is_subtype` function):

```python
from typing import Sequence, List, Union, Optional, Any

def is_subtype(hintA, hintB) -> bool:
    ...  # ?

assert is_subtype(list, list) is True
assert is_subtype(list, Sequence) is True
assert is_subtype(Sequence, list) is False

assert is_subtype(List[int], Sequence[Any]) is True
assert is_subtype(List[int], Sequence[int]) is True
assert is_subtype(List[int], Sequence[str]) is False

assert is_subtype(list, Union[list, str]) is True
assert is_subtype(Union[int, str], Union[int, str, list]) is True
assert is_subtype(Union[list, str], list) is False
assert is_subtype(list, Union[int, str]) is False
assert is_subtype(Union[int, tuple], Union[int, str, list]) is False

assert is_subtype(Optional[int], int) is False
assert is_subtype(int, Optional[int]) is True
```

is this something that could be achieved with `beartype`?  digging through the source, i see _lots_ of stuff that seems very useful towards this goal, but not sure I see anything in the public API (since we're not really trying to determine whether some object is an instance of a hint).

my goal here would be to detect if an API breakage has occurred between two versions:

```py
# version 1
def func() -> list:
   ...
```

```py
# version 2
def func() -> Sequence:  # potential API breakage
   ...
```

... and this seems like something you'd have an opinion on :) 
thanks for your work!

**Comment by leycec** [comment]
**Fascinating.** Something spicey resembling `beartype.is_subtype()` is *totally* in @beartype's wheelhouse. Sadly, as you surmise, @beartype currently lacks a public API for testing type hint compatibility via (super|sub)set theoretic operators like ⊆ and ⊂ implemented through Python's builtin `in` operator – mostly because you're the first fearless bear-bro to ever request this.

> my goal here would be to detect if an API breakage has occurred between two versions:

Yes! Thanks for that perfect use case as well. The real-world applicability is clear. If you've typed an existing callable as returning a type hint `A`, can you safely re-type that same callable as returning a type hint `B` while preserving backward compatibility? Equivalently, in set theoretic notation, is `A ⊆ B`?

So let's suppose we define this fun new `beartype.is_subtype()` API as we eventually should. I'm envisioning that function mostly being called interactively from Python REPLs like IPython3 or Jupyter Notebook. Is that right? Okay. So...

## Let's Just Pretend Type Hints Are Sets, Guys

Scarce volunteer resources rears its ugly mullet cut yet again. Briefly inspecting [the `pytypes.is_subtype()` tester](https://github.com/Stewori/pytypes/blob/3846f1be3e44892c8e60f750f7581cbddd8be383/pytypes/type_util.py#L1812) suggests this problem domain to be *really, really* non-trivial. That's probably the most significant blocker here.

It sort of makes sense this would be the NP-hard of type hints, right? After all, type hints *aren't* sets. They're type hints. There's nothing innately set-like about type hints. So, attempting to coerce type hints into sets for the purpose of deciding superset and subset containment relations between otherwise unrelated type hints rapidly descends into some fairly intense academic theory-crafting.

Of course, we're all for theory-crafting here. @beartype is already built on a flimsy foundation of hand-waving, wishful thinking, and helpful assumptions that mostly hold up until you squint at them. Still... `pytypes` is dead. I can't help but think that part of the reason `pytypes` died is exactly this: `pytypes` authors devoted a shocking slice of their scarce time to grotesquely NP-hard-like problems that *most* of their userbase didn't particularly care about.

Let's not throw ourselves like disgruntled lemmings into that same hole. :hole: 

## So What Are You Saying Here, @leycec?

Welp... I'm *kinda* sayin' this may never happen. I mean, I really hope it does! But it's a surprisingly hardcore feature that would [strongly benefit from funding sponsorship of some sort](https://github.com/sponsors/leycec). Lucrative grant funding agencies across the industrial world: make this happen for all humanity, please.

Let's see what the unruly rabble says. Does anyone else *really* want this too – or can I glibly pretend this never happened and devote myself to other h0t ticket dumpster fires that are currently on fire?

**Reply by tlambert03** [reply]
lolll ... I commend your honesty :joy:

I am not surprised to hear you say that.  I kinda supposed it was a hardcore feature.  I'll promise to try to sneak your name into any conversations I happen to have with deep-pocketed FOSS lovers 🤑

**however**... in the meantime, let's suppose I'm a positively gruntled lemming who's a bit obsessed with this problem at the moment, likely a bit naïve, and that you're not going to lose me that easily.  (but that I also definitely don't expect to pull you so easily from your many other dumpster fires which we can all understand and empathize with).

And let's also suppose that I'm ok with thinking about this in stages.   If you look at the pytypes implementation, for all its complexity and indirection, it boils down to this:

```py
from typing import ForwardRef, TypeVar

def is_subtype(sub, sup):
    if isinstance(sub, ForwardRef) or isinstance(sup, ForwardRef): ...
        # deal with all the nastiness of ForwardRefs
    if issubclass(sub, Empty) or issubclass(sup, Empty): ...
        # where Empty is a special pytypes class trying to deal with
        # https://github.com/python/typing/issues/157
    if isinstance(sup, TypeVar) or isinstance(sub, TypeVar): ...
        # deal with __bound__, __contravariant__, __covariant__
    if is_Tuple(sup): ...
    if is_Union(sup): ...
    if is_Union(sub): ...
    if is_Generic(sup): ...
    try:
        return issubclass(sub, sup)
    except TypeError:
        return False
```

- ForwardRefs are annoying, but that's a different (namespace) problem.
- I'm ok with ignoring the [`Empty`](https://github.com/python/typing/issues/157) case for now, maybe forever.
- TypeVar is bounds and co/contra-variants are hard, but I'd happily raise a `NotImplementedError` for now and likely still profit quite a bit.
- after that, we're down to Generics – probably the bulk of the "non-standard" work
- then the rest is "straightforward" enough.

If I _were_ to dig into this some more on my own time – likely using some of your private API and fully expecting it to break without notice – would you have any additional "here's where I might start if I were you" sort of thoughts?  (s'ok if not!)  I feel like there's some gold buried in `sanify_hint_root` and in `_util/hint/*` that could get this started off on at least a nice normalized footing.
 
thanks already for your time and thorough response!

**Comment by leycec** [comment]
**Yes, yes, and yes.** While clearing the ruinous fallout of a 150 year-old oak tree that fell over during [the deadliest derecho in Ontario history](https://www.cbc.ca/news/canada/toronto/ontario-storm-climate-change-election-1.6463877), crushed our bunkhouse, and then toppled onto our cottage, I had a miraculous epiphany. Praise Guido!

I didn't *intend* to think about this – but there I was, hauling debris, petting cats, and thinking about this despite my best intentions. But first...

## What Not to Do

What not to do is what `pytypes` does – probably. You've spelunked deeper into the cavernous labyrinth of the `pytypes` codebase than I have. So, this is mere baseless speculation on my part. But...

What `pytypes` does is guaranteed to fail – probably. If I'm reading the `pytypes.type_util` submodule correctly, the `pytypes.is_subtype()` function *basically* reduces to a linear chain of `if` conditionals. Your superb synopsis of that function corroborates that I am not entirely insane. Sadly, a linear chain of `if` conditionals can't possibly suffice to cover all possible edge cases. Why? Because type hints are infinitely composable (i.e., recursively nestable). For example, consider just this simple fragment from your `pytypes.is_subtype()` simplification:

```python
    if is_Tuple(sup): ...
    if is_Union(sup): ...
    if is_Union(sub): ...
    if is_Generic(sup): ...
```

So, what if both the subtype and supertype are type hints containing multiple generics subscripting multiple tuples subscripting multiple unions subscripting a dictionary subscripting a list (e.g., `list[dict[Union[tuple[MuhGeneric, MuhOtherGeneric, ...], int], str]]`)? Those are hardly worst-case type hints, either.

Since there are a countably infinite number of possible type hints, there's an even *larger* number of 2-combinations `(subtype, supertype)` for any arbitrary type hint `subtype` being compared against any arbitrary type hint `supertype`. Ergo, there is some 2-combination of type hints that *cannot* possibly be handled by a linear chain of `if` conditionals. Indeed, the binomial coefficient tells us that the number of possible 2-combinations between `n` type hints grows explosively as:

```
(n 2) = "n pick 2" = n!/[2(n-2)!]
```

This is why C++ popularized operator overloading, right? By overloading the `<=` comparator within four C++ classes `A`, `B`, `C`, and `D` in a compatible manner, any two instances of those four classes can be implicitly compared **without having to explicitly code a linear chain of `if` conditionals.** That's good! Explicitly coding a linear chain of `if` conditionals rapidly becomes impossible in practice, because [combinatorial explosion makes my brain burn](https://en.wikipedia.org/wiki/Combinatorial_explosion).

## What to Do, Then?

What to do is what we *always* do in Computer Science when confronted with a seemingly insurmountable, intractable, and infeasible mountain of pain: we transform that complex problem we cannot solve into a simpler problem we have already solved.

In this case, Python has already solved a simpler problem: **deciding whether one type is a subclass of another** (i.e., the `issubclass(A, B)` relation). If we can transform our complex `is_subtype(A, B)` relation we cannot <sup>readily</sup> solve into the simpler `issubclass(A, B)` relation that Python has already solved, then our job is done before it even began. Can we do this?

**Yes, we can.** <sup>*you knew that was coming*</sup> B-b-but... how can we? "Simple." The `beartype.is_subtype(subtype, supertype)` tester function should probably:

1. Dynamically create one in-memory class for each inclusive superclass of each nested type hint of each passed type hint, ultimately producing two internal type hierarchies: a type hierarchy `SubType` representing the passed `subtype` type hint and yet another type hierarchy `SuperType` representing the passed `supertype` type hint. This operation has `O(jk)` time complexity (I think?), where:
   * `j` is the maximum number of superclasses of any nested type hint, equivalent to the length of the method resolution order (MRO) for that hint (i.e., `j = len(type_hint.__mro__)`).
   * `k` is the maximum number of nested type hints of each passed type hint.
2. Return the result of calling `issubclass(SubType, SuperType)`.

Obviously, **2.** is a trivial one-liner. That just leaves **1.**, which is non-trivial but tractable. Consider this concrete example:

```python
>>> from beartype import is_subtype
>>> from beartype.typing import list, Sequence, Union

# User-defined "int" subclass, just 'cause.
>>> class MuhInt(int): pass

>>> subtype = list[MuhInt]
>>> supertype = Union[Sequence[int], str]
>>> is_subtype(subtype, supertype)
True
```

Internally, `beartype.is_subtype()` will create two type hierarchies: one representing the passed subtype and the other representing the passed supertype. For this concrete example, these type hierarchies might resemble:

```python
     # v--- beartype-specific fast caching protocols for the win
from beartype.typing import Protocol

# ------[ TYPE HIERARCHY FOR: list[MuhInt] ]------
# Type representing the type hint "list[object]".
class ListOfObject(list, Protocol): pass

# Type representing the type hint "list[int]".
class ListOfInt(ListOfObject, Protocol):
    def has_int(): pass

# Type representing the type hint "list[MuhInt]".
class ListOfMuhInt(ListOfInt, Protocol):
    def has_MuhInt(): pass

# ------[ TYPE HIERARCHY FOR: Union[Sequence[int], str] ]------
# Type representing the type hint "Sequence[object]".
from collections.abc import Sequence
class SequenceOfObject(Sequence, Protocol): pass

# Type representing the type hint "Sequence[int]".
class SequenceOfInt(SequenceOfObject, Protocol):
    def has_int(): pass

# Type representing the type hint "Union[Sequence[int], str]".
UnionOfSequenceOfIntAndStr = SequenceOfInt | str
         # ^--- check dat dark PEP 604 magic out, bro
```

We can both agree there's a tractable algorithm for producing those two type hierarchies, right? Right! You're darn right! Indeed, note that Python itself already provides an out-of-the-box comparison between the builtin `list` type and the `collections.abc.Sequence` ABC:

```python
>>> from collections.abc import Sequence
>>> issubclass(list, Sequence)
True
```

Because both the `ListOfMuhInt` and `SequenceOfInt` protocols declare the same `has_int()` method, the former is necessarily a subclass of the latter under [PEP 544-style structural subtyping](https://peps.python.org/pep-0544/). :open_mouth: 

We can also then agree that we've transformed our complex `is_subtype(A, B)` relation we cannot <sup>readily</sup> solve into the simpler `issubclass(A, B)` relation that Python has already solved. Altogether, this means that:

```python
# This non-trivial function call copied from above...
>>> is_subtype(list[MuhInt], Union[Sequence[int], str])
# ...is now trivially reducible to this obvious function call
# leveraging the two type hierarchies we generated above.
>>> issubclass(ListOfMuhInt, UnionOfSequenceOfIntAndStr)
True
```

I've tested that. It actually works. Shocked Pikachu face, arise!

![oh boy](https://en.meming.world/images/en/thumb/2/2c/Surprised_Pikachu_HD.jpg/600px-Surprised_Pikachu_HD.jpg)

## ...Can't Be That Easy

*...heh.* Yeah. You just know something's gonna blow that trivial algorithm up hard. But until that happens, let's choose to believe in the miraculous healing power of Python's `issubclass()` builtin. :grin:

**Comment by leycec** [comment]
Also, I hope you don't mind – but [I've promoted this to a full-blown feature request on our issue tracker](https://github.com/beartype/beartype/issues/133)! If you wouldn't mind, let's continue blathering over there so I can keep track of this shambling monstrosity.
